### DISCUSSION TITLE
libzmq.so.5 library not found

### DISCUSSION THREAD

**Discussion by wangqingpen** [edgexfoundry/community#205] [Q&A]
docker compilation has been completed but docker keeps restarting, the log is as follows![image](https://github.com/edgexfoundry/community/assets/72652818/185432ea-4fec-43b0-92a6-ef7b4b633ce2)
Shows missing .so file, but I've installed it in dockerfile, dockerflie file as follows
![image](https://github.com/edgexfoundry/community/assets/72652818/d3eb50f0-2302-486b-b209-ddde60c01277)

**Comment by bnevis-i** [comment]
It definitely looks like a link library loading issue.  Are you using c linkage when compiling the go binary?

Is this your own code, or something on github.com/edgexfoundry?  What tag/branch/component is failing?

**Reply by wangqingpen** [reply]
Yes, I rewrite modbus  But I've put the link library in ，Pls is there any way to fix it thanks

**Reply by bnevis-i** [reply]
Can you share the full contents of the Dockerfile and Makefile that you are using?

**Reply by wangqingpen** [reply]
**makefile**
.PHONY: build test clean docker

export GODEBUG=cgocheck=0



GO=CGO_ENABLED=1 GO111MODULE=on GOPROXY=https://goproxy.io,direct go
#GOCGO=CGO_ENABLED=1 GO111MODULE=on GOPROXY=https://goproxy.cn,direct go

MICROSERVICES=cmd/device-dlt698.45-go
.PHONY: $(MICROSERVICES)

VERSION=$(shell cat ./VERSION 2>/dev/null || echo 1.3.1)
DOCKER_TAG=$(VERSION)-dev

GOFLAGS=-ldflags "-X github.com/edgexfoundry/device-dlt698.45-go.Version=$(VERSION)"
GOTESTFLAGS?=-race

GIT_SHA=$(shell git rev-parse HEAD)

build: $(MICROSERVICES) -lm
	$(GO) install -tags=safe

cmd/device-dlt698.45-go:
	$(GO) build $(GOFLAGS) -o $@ ./cmd/device-dlt698.45-go

docker:
	docker build \
		-f Dockerfile \
		--label "git_sha=$(GIT_SHA)" \
		-t edgexfoundry/docker-device-dlt698.45-go:1.3.1 \
		.
docker-arm64:
	docker build \
		-f cmd/device-dlt698.45-go/Dockerfile-arm \
		--label "git_sha=$(GIT_SHA)" \
		-t edgexfoundry/docker-device-dlt698.45-go-arm64:1.3.1 \
		.
docker-armhf:
	docker build \
		-f cmd/device-dlt698.45-go/Dockerfile-arm \
		--label "git_sha=$(GIT_SHA)" \
		-t edgexfoundry/docker-device-dlt698.45-go-armhf:1.3.1 \
		.
test:
	GO111MODULE=on go test $(GOTESTFLAGS) -coverprofile=coverage.out ./...
	GO111MODULE=on go vet ./...
	gofmt -l .
	[ "`gofmt -l .`" = "" ]
	./bin/test-attribution-txt.sh
	./bin/test-go-mod-tidy.sh

clean:
	rm -f $(MICROSERVICES)
**dockerfile**  
 ARG BASE=golang:1.17-alpine3.12
FROM ${BASE} AS builder
RUN apt-get update
RUN apt-get install make git gcc libc-dev   -y

FROM golang:1.17-alpine AS builder
ARG MAKE='make build'
ARG ALPINE_PKG_BASE="make git  tzdata openssh-client gcc libc-dev zeromq-dev libzmq3-dev libsodium-dev"
WORKDIR /device-dlt698.45-go


LABEL license='SPDX-License-Identifier: Apache-2.0' \
  copyright='Copyright (c) 2020: Intel'



RUN sed -e 's/dl-cdn[.]alpinelinux.org/nl.alpinelinux.org/g' -i~ /etc/apk/repositories

# add git for go modules


RUN apk add --update --no-cache ${ALPINE_PKG_BASE} \
&& cp /usr/share/zoneinfo/Asia/Shanghai /etc/localtime \
&& echo "Asia/Shanghai" > /etc/timezone


ENV GOPROXY="https://goproxy.cn,direct"
#ENV DEBIAN_FRONTEND noninteractive

COPY go.mod .
# add git for go modules
RUN apk add --update --no-cache make git

RUN go mod download

COPY . .
RUN [ ! -d "vendor" ] && go mod download all || echo "skipping..."


WORKDIR /device-dlt698.45-go
RUN apk info -a zeromq

RUN ${MAKE}

# Next image - Copy built Go binary into new workspace
FROM alpine:3.12
ENV GODEBUG=cgocheck=0

ENV GOPROXY="https://goproxy.cn,direct"
#ENV GOPROXY="https://goproxy.cn,direct"

#RUN sed -e 's/dl-cdn[.]alpinelinux.org/nl.alpinelinux.org/g' -i~ /etc/apk/repositories
RUN sed -e 's/dl-cdn.alpinelinux.org/mirrors.ustc.edu.cn/g' /etc/apk/repositories
RUN apk add --update --no-cache libzmq3-dev

ENV LD_LIBRARY_PATH=/usr/local/bin
LABEL license='SPDX-License-Identifier: Apache-2.0' \
  copyright='Copyright (c) 2020: Intel'

ENV APP_PORT=59930
#expose command data port
EXPOSE $APP_PORT

WORKDIR /
COPY --from=builder device-dlt698.45-go/lib /lib
COPY --from=builder device-dlt698.45-go/cmd/device-dlt698.45-go/device-dlt698.45-go /usr/local/bin/device-dlt698.45-go
#COPY --from=builder device-dlt698.45-go/cmd/device-dlt698.45-go/device-dlt698.45-go /device-dlt698.45-go
COPY --from=builder device-dlt698.45-go/cmd/device-dlt698.45-go/res/configuration.toml /res/configuration.toml
COPY --from=builder device-dlt698.45-go/cmd/device-dlt698.45-go/res/license.dat /res/license.dat

#VOLUME ["/res/"]

ENTRYPOINT ["/usr/local/bin/device-dlt698.45-go"]
CMD ["-cp=consul.http://edgex-core-consul:8500", "--registry", "--confdir=/res"]

**Reply by bnevis-i** [reply]
Lastly, what version of the go device SDK are you using?

**Reply by wangqingpen** [reply]
v2

**Comment by jiekechoo** [comment]
Please try run docker container manully: `docker run -it --rm  --entrypoint sh edgexfoundry/docker-device-dlt698.45-go:1.3.1`, and use command: `find / -name libzmq.*` check `libzmq` has installed.

**Reply by wangqingpen** [reply]
![image](https://github.com/edgexfoundry/community/assets/72652818/0e7d34d6-1652-474e-92e8-9085e321bf62)
already installed

**Reply by jiekechoo** [reply]
Can you `make build` in your develop host, and run it manully? Paste all build progress and run log here.

**Reply by wangqingpen** [reply]
**make build** 
![image](https://github.com/edgexfoundry/community/assets/72652818/b50fae72-7f10-4f04-94b4-7e0434ac9b99)
**logs**
![image](https://github.com/edgexfoundry/community/assets/72652818/d104f2cb-5dfd-4cdc-b385-f9f4c285930f)

**Reply by wangqingpen** [reply]
This is normal operation

**Comment by bnevis-i** [comment]
One thing that seems suspicious to me is the number of FROM lines in the Dockerfile:

```
ARG BASE=golang:1.17-alpine3.12
FROM ${BASE} AS builder
...
FROM golang:1.17-alpine AS builder
...
FROM alpine:3.12
```

It is critical that the build container uses the same base version of alpine as the runtime container.  Here, the middle FROM statement doesn't specifically lock down the alpine version.

It looks like it is looking for libzmq.so.5 in both cases, and it is unlikely there would be a breaking change without rolling the major version.  But, I thought I would mention the anomaly.

**Reply by wangqingpen** [reply]
ARG BASE=golang:1.17-alpine3.15
FROM ${BASE} AS builder

FROM alpine:3.12   I've changed it to the above command, and it still gets the same error

**Comment by lenny-goodell** [comment]
@wangqingpen , please upgrade to EdgeX 3.1 (just released). We have removed ZMQ as a dependency as of EdgeX 3.0.
https://wiki.edgexfoundry.org/display/FA/Napa

**Reply by wangqingpen** [reply]
The rest of my parts are v2 versions. Is V3 compatible with V2? For example, services are often displayed on the UI

**Reply by cloudxxx8** [reply]
No, there are some breaking changes in v3, so they are not compatible.

**Reply by wangqingpen** [reply]
This is too much work, is there any other workaround

**Reply by wangqingpen** [reply]
Because my other services are all v2 versions

**Reply by cloudxxx8** [reply]
You can remove the zero mq from the Docker file and rebuild all the go binary with `no_zmq` build tag

**Reply by cloudxxx8** [reply]
https://github.com/edgexfoundry/edgex-go/blob/4f1b2626cb078c7149dc655301029b2517b4b500/Makefile#L83

**Reply by wangqingpen** [reply]
You mean to remove zmq and put NO_ZMQ_GO_BUILD_TAG:=no_zmq into the dockerfile?

**Reply by cloudxxx8** [reply]
remove zmq from Docker file
put no_zmq to your go build command
